src/section/shared_state.rs

```diff
@@ -52,25 +55,28 @@ impl SharedState {
         }
     }
 
-    pub fn update(&mut self, new: Self) {
+    pub fn update(&mut self, new: Self) -> Result<(), RoutingError> {
         if self.handled_genesis_event {
-            log_or_panic!(
-                log::Level::Error,
-                "shared state update - genesis event already handled",
-            );
+            error!("shared state update - genesis event already handled",);
+            return Err(RoutingError::InvalidState);
         }
 
-        if self.our_history.len() > 1 && *self != new {
-            log_or_panic!(
-                log::Level::Error,
-                "shared state update - mismatch: old: {:?} --- new: {:?}",
-                self,
-                new
-            );
+        if self.our_history.len() > 1 {
+            if *self != new {
+                error!(
+                    "shared state update - mismatch: old: {:?} --- new: {:?}",
+                    self, new
+                );
+                return Err(RoutingError::InvalidState);
+            }
+        } else if !self.our_history.self_verify() {
```

Comment:
Two points:
1. You should do the check on `new`, not `self`. `self` is presumably valid, because we only allow approved mutations. But we don't know whether `new` is, because we just received it from someone.
2. You should verify everything, not just `our_history`.

Reply:
> You should do the check on new

Ah, right. 

> You should verify everything, not just our_history

That means a check against the serialised info itself? I thought such check shall be carried out inside the message accumulation already?
Regarding checking each field individually, I am not sure for each field, which value is `valid`.
For example, `handled_genesis_event` shall be `false` or `true` for new? and `our_members` or `relocate_queue` shall be what? etc.

Reply:
> I thought such check shall be carried out inside the message accumulation already

No, the idea is that any node can send us a message containing a `SharedState` and we should be able to merge it with outs if it is valid. There will be no accumulation for this, because all the info contained in `SharedState` is already signed by the section quorum. Currently this isn't necessary, because we only call `update` in response to accumulated `Genesis` event, but that will eventually change.

Reply:
> For example, handled_genesis_event shall be false or true for new? and our_members or relocate_queue shall be what? etc.

Right, so not everything is ready for parsec removal yet, but we can do at least those parts that are. So:
- `handled_genesis_event` can be ignored (it will always be `false` for incoming shared state)
- `relocate_queue` can be ignored too, for now, at least until we figure out what to do with it (we have a separate issue for it and I think you started looking into it?)
- `our_members` can be verified normally, because the entries in it (`MemberInfo`) all have proofs attached.
- Same with `SectionMap`

Reply:
Ok, will do the checks.

>it will always be false for incoming shared state

Then maybe just check the incoming `new`'s `handled_genesis_event ` shall be `false` ?